lightning/src/chain/chainmonitor.rs

```diff
@@ -0,0 +1,226 @@
+	/// block to obtain updated `txdata` and re-call `block_connected`.
+	///
+	/// [`ChannelMonitor::block_connected`]: ../channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`chain::Watch::release_pending_monitor_events`]: ../trait.Watch.html#tymethod.release_pending_monitor_events
+	/// [`chain::Filter`]: ../trait.Filter.html
+	pub fn block_connected(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) -> bool {
+		let mut has_new_outputs_to_watch = false;
+		{
+			let mut monitors = self.monitors.lock().unwrap();
+			for monitor in monitors.values_mut() {
+				let mut txn_outputs = monitor.block_connected(header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+				has_new_outputs_to_watch |= !txn_outputs.is_empty();
+
+				if let Some(ref chain_source) = self.chain_source {
+					for (txid, outputs) in txn_outputs.drain(..) {
+						for (idx, output) in outputs.iter().enumerate() {
+							chain_source.register_output(&OutPoint { txid, index: idx as u16 }, &output.script_pubkey);
+						}
+					}
+				}
+			}
+		}
+		has_new_outputs_to_watch
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel based on the disconnected block. See [`ChannelMonitor::block_disconnected`] for
+	/// details.
+	///
+	/// [`ChannelMonitor::block_disconnected`]: ../channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	pub fn block_disconnected(&self, header: &BlockHeader, disconnected_height: u32) {
+		let mut monitors = self.monitors.lock().unwrap();
+		for monitor in monitors.values_mut() {
+			monitor.block_disconnected(header, disconnected_height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+		}
+	}
+
+	/// Creates a new `ChainMonitor` used to watch on-chain activity pertaining to channels.
+	///
+	/// When an optional chain source implementing [`chain::Filter`] is provided, the chain monitor
+	/// will call back to it indicating transactions and outputs of interest. This allows clients to
+	/// pre-filter blocks or only fetch blocks matching a compact filter. Otherwise, clients may
+	/// always need to fetch full blocks absent another means for determining which blocks contain
+	/// transactions relevant to the watched channels.
+	///
+	/// [`chain::Filter`]: ../trait.Filter.html
+	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F) -> Self {
+		Self {
+			monitors: Mutex::new(HashMap::new()),
+			chain_source,
+			broadcaster,
+			logger,
+			fee_estimator: feeest,
+		}
+	}
+
+	/// Adds the monitor that watches the channel referred to by the given outpoint.
+	///
+	/// Calls back to [`chain::Filter`] with the funding transaction and outputs to watch.
+	///
+	/// [`chain::Filter`]: ../trait.Filter.html
+	fn add_monitor(&self, outpoint: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), MonitorUpdateError> {
```

Comment:
This and update_monitor below can just be inlined in the `impl chain::Watch` block, since they're non-pub and only called from there.

Reply:
Oh, maybe dont do this, since Val did it in 681 and no need to make her rebase.

Reply:
Yep, was thinking the same. @valentinewallace